### PR TITLE
Added the ability to upload honk viruses to doors, machines and computers. Added new sound_player component.

### PR DESCRIPTION
**Pull request by KathyRyals**
## About The Pull Request

You can use your PDA with a Honkworks 5.0 cartridge in it to upload a virus that will make an airlock, machine or computer honk for a while.
I also made a new generic sound_player component to be used as a way to play any sound(s) when receiving any signal(s).

## Why It's Good For The Game

It's funny behaviour and cool flavour.
Also good code improvement with robust, generic component.

## TODO

- [x] Refactor to use components.

- [x] Virus uploading to every machine with a panel open / computer.

- [x] Making a generic sound_player component to handle any sound playing with any signals.

## Changelog
:cl: Rohesie and Kathy Ryals
add: The Honkworks 5.0 cartridge's software has been updated to be able to hack machines, computers and airlocks to make them honk for a while.
code: Added a new sound_player component to be able to play any sound with any signals.
/:cl:

### CODE REVIEW

**Comment by NanoCats** [comment]
Make it so that you don't need to be wearing the shoes to slip people while we're at it, honk.

**Comment by KathyRyals** [comment]
I will do it, and use an headmin ruling as justification.

**Hulkamania**
Shoeless clowns have been specifically encouraged via code changes to allow them to wear appropriate clothing for dress up shenanigans. No longer do they have a negative moodlet for being without their shoes. A clown is supposed to dress up to their hearts content, limiting that by making them valid without the shoes is rather punishing.

**Comment by ATH1909** [comment]
image

That thing that NanoCats suggested is basically a straight revert of https://github.com/tgstation/tgstation/pull/50650. If you remove the clown shoes requirement from clown PDA slipping, gamers (and dishonorable clowns) will wear clown PDAs (without shoes) just for the ability to slip sec officers while knocked down from a baton strike (or for the ability to lay down and sometimes slip your captor while cuffed).

**Comment by NanoCats** [comment]
> That thing that NanoCats suggested is basically a straight revert of #50650. If you remove the clown shoes requirement from clown PDA slipping, gamers (and dishonorable clowns) will wear clown PDAs (without shoes) just for the ability to slip sec officers while knocked down from a baton strike (or for the ability to lay down and sometimes slip your captor while cuffed).

I mean, again, clowns are meant to be able to dress up for fun, hm. I get what you're saying, but uh, I dunno, maybe make it so the HONKworks cartridge is what grants the slip ability instead of just the clown PdA?

I think this isn't that bad of an exploit sicne to do so you have to get the HOP to repaint your PDA to clown, or break in.

**Comment by ATH1909** [comment]
> > That thing that NanoCats suggested is basically a straight revert of #50650. If you remove the clown shoes requirement from clown PDA slipping, gamers (and dishonorable clowns) will wear clown PDAs (without shoes) just for the ability to slip sec officers while knocked down from a baton strike (or for the ability to lay down and sometimes slip your captor while cuffed).
> 
> I mean, again, clowns are meant to be able to dress up for fun, hm. I get what you're saying, but uh, I dunno, maybe make it so the HONKworks cartridge is what grants the slip ability instead of just the clown PdA?
> 
> I think this isn't that bad of an exploit sicne to do so you have to get the HOP to repaint your PDA to clown, or break in.

People are going to specifically gank the clown so that they can nab his PDA, just as they gank the captain or HoP for their all access cards or the CE for his boots.

**Comment by NanoCats** [comment]
> > > That thing that NanoCats suggested is basically a straight revert of #50650. If you remove the clown shoes requirement from clown PDA slipping, gamers (and dishonorable clowns) will wear clown PDAs (without shoes) just for the ability to slip sec officers while knocked down from a baton strike (or for the ability to lay down and sometimes slip your captor while cuffed).
> > 
> > 
> > I mean, again, clowns are meant to be able to dress up for fun, hm. I get what you're saying, but uh, I dunno, maybe make it so the HONKworks cartridge is what grants the slip ability instead of just the clown PdA?
> > I think this isn't that bad of an exploit sicne to do so you have to get the HOP to repaint your PDA to clown, or break in.
> 
> People are going to specifically gank the clown so that they can nab his PDA, just as they gank the captain or HoP for their all access cards or the CE for his boots.

I mean, people do those things, people valid clowns for not wearing their shoes on servers that aren't manuel, I just don't think this is the biggest of deals. Cluwnes should be able to dress up without having to sacrifice their slippy magic (ESPECIALLY because people can just jack your shoes)

**Comment by ATH1909** [comment]
But hey, maybe I'm just overreacting. I didn't see too many antags rush clown PDAs before the clown shoes restriction was added...

I guess what I'm saying is that the clown shoes restriction removal should be in its own PR, not bundled with a genuinely good feature like the ability to honk-hack doors.

**Comment by ArcaneDefence** [comment]
>pr about a silly interaction that clown can use to make better use out of the honk virus

>suddenly buffs the shit out of clown combat


Where did we go so wrong?

**Comment by Iatots** [comment]
document the changelog too

**Comment by KathyRyals** [comment]
I can take it out no skin off my back, I didn't know it was such a big deal !

**Comment by KathyRyals** [comment]
No more shoe thing. Only virus now.

**Comment by GuyonBroadway** [comment]
Clowns who replace their shoes with a pair of sneakers at roundstart deserve a long walk out a short airlivk, don't  @ me.

**Comment by Rohesie** [comment]
I can see what you are doing here, and the feature is very nice, but this needs to be done through signals.
Poke us on the coder channel if you want to learn how.

**Comment by itseasytosee** [comment]
My honor is defended thank you for your consideration and reason.

File: code/datums/components/doorhonk.dm [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	var/honkSound = 'sound/items/bikehorn.ogg'
+	var/volume = 30
+	var/honkLeft
+	var/infinite = FALSE
+
+/datum/component/doorhonk/Initialize(volume_override = 30, infinite_honking = FALSE)
+	if(!istype(parent, /obj/machinery/door/airlock)) //This component only works for doors.
+		return COMPONENT_INCOMPATIBLE
+
+	honkLeft = (rand(15,20)) //This component works for an amount of honk between 15 and 20
```
Comment (Rohesie) [inline code/datums/components/doorhonk.dm:12]: Include this in the parameters. Something like this:
`/datum/component/doorhonk/Initialize(volume_override = 30, honk_amount = rand(15, 20))`
Then you can add something like `if(honk_amount <= 0) infinite = TRUE`

File: code/game/machinery/doors/airlock.dm [hunk 2]
```diff
@@ -1037,6 +1037,8 @@
 			return FALSE
 		use_power(50)
 		playsound(src, doorOpen, 30, TRUE)
+		//TODO SEND_SIGNAL
+		SEND_SIGNAL(src, COMSIG_AIRLOCK_OPEN, )
```
Comment (Rohesie) [inline code/game/machinery/doors/airlock.dm:1041]:
```suggestion
		SEND_SIGNAL(src, COMSIG_AIRLOCK_OPEN, forced)
```
Always pass the args of the proc

File: code/game/machinery/doors/airlock.dm [hunk 3]
```diff
@@ -1085,6 +1087,7 @@
 			return
 		use_power(50)
 		playsound(src, doorClose, 30, TRUE)
+		SEND_SIGNAL(src, COMSIG_AIRLOCK_CLOSE)
```
Comment (Rohesie) [inline code/game/machinery/doors/airlock.dm:1090]:
```suggestion
		SEND_SIGNAL(src, COMSIG_AIRLOCK_CLOSE, forced)
```

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 4]
```diff
@@ -979,6 +979,22 @@ GLOBAL_LIST_EMPTY(PDAs)
 	. = ..()
 	if(!proximity)
 		return
+	if(istype(A,/obj/machinery/door/airlock))
+		if(!istype(cartridge, /obj/item/cartridge/virus/clown))
+			return
```
Comment (Rohesie) [inline code/game/objects/items/devices/PDA/PDA.dm:984]: Use either `pre_attack` or `attackby`

File: code/game/machinery/doors/airlock.dm (hunk 2 above)
Comment (Rohesie) [reply code/game/machinery/doors/airlock.dm:1041]: Also, place this signal right after `operating = TRUE`
We want to send it right before the event happens. The condition for a door to begin opening, without a chance to go back, is when that happens.

File: code/game/machinery/doors/airlock.dm (hunk 3 above)
Comment (Rohesie) [reply code/game/machinery/doors/airlock.dm:1090]: Same here, place it right over `operating = TRUE`

File: code/datums/components/doorhonk.dm (hunk 1 above)
Comment (KathyRyals) [reply code/datums/components/doorhonk.dm:12]: Doesn't work, can't call rand in parameters.

**Comment by Asha108** [comment]
Looks great. Can't wait to honk all the machines on the station and to get lynched for it.

File: code/__DEFINES/dcs/signals.dm [hunk 5]
```diff
@@ -429,6 +429,13 @@
 ///from base power_change() when power is restored
 #define COMSIG_MACHINERY_POWER_RESTORED "machinery_power_restored"
 
+// /obj/machinery/door/airlock signals
+
+//from /obj/machinery/door/airlock/open(forced=0)
+#define COMSIG_AIRLOCK_OPEN "airlock_open"
+//from /obj/machinery/door/airlock/close(forced=0)
```
Comment (Rohesie) [inline code/__DEFINES/dcs/signals.dm:436]:
```suggestion
//from /obj/machinery/door/airlock/open(): (forced)
#define COMSIG_AIRLOCK_OPEN "airlock_open"
//from /obj/machinery/door/airlock/close(): (forced)
```

File: code/game/machinery/doors/airlock.dm [hunk 6]
```diff
@@ -1029,6 +1029,7 @@
 /obj/machinery/door/airlock/open(forced=0)
 	if( operating || welded || locked )
 		return FALSE
+	SEND_SIGNAL(src, COMSIG_AIRLOCK_OPEN, forced)
```
Comment (Rohesie) [inline code/game/machinery/doors/airlock.dm:1032]: Put this right over `operating = TRUE`, which is when the change happens.

File: code/game/machinery/doors/airlock.dm [hunk 7]
```diff
@@ -1069,6 +1071,7 @@
 /obj/machinery/door/airlock/close(forced=0)
 	if(operating || welded || locked)
 		return
+	SEND_SIGNAL(src, COMSIG_AIRLOCK_CLOSE, forced)
```
Comment (Rohesie) [inline code/game/machinery/doors/airlock.dm:1074]: Put this right over `operating = TRUE`, which is when the change happens.

**Comment by KathyRyals** [comment]
Tested locally and working !

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 8]
```diff
@@ -904,6 +904,29 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+		var/obj/machinery/target_machine = target
+		if(!target_machine.panel_open && !istype(target, /obj/machinery/computer))
+			return ..()
+		if(!istype(cartridge, /obj/item/cartridge/virus/clown))
+			return ..()
+		var/obj/item/cartridge/virus/installed_cartridge = cartridge
+
+		if(installed_cartridge.charges <=0)
+			to_chat(user, "<span class='notice'>Out of charges.</span>")
+			return ..()
```
Comment (Rohesie) [inline code/game/objects/items/devices/PDA/PDA.dm:918]: Say you want to use your PDA with an ID in it to open a door, like you can do normally. Will you always get this message then?

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 9]
```diff
@@ -904,6 +904,29 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+		to_chat(user, "<span class='notice'>You upload the virus to the airlock controller!</span>")
+		var/sig_list
+		if(istype(target,/obj/machinery/door/airlock))
+			sig_list += list(COMSIG_AIRLOCK_OPEN, COMSIG_AIRLOCK_CLOSE)
+		else
+			sig_list += list(COMSIG_ATOM_ATTACK_HAND)
+			//TODO : Add more signals for specific cases like panel open, printing, dispensing etc
+		target.AddComponent(/datum/component/sound_player, amount = (rand(15,20)), signal_or_sig_list = sig_list)
+		installed_cartridge.charges --
+
```
Comment (Rohesie) [inline code/game/objects/items/devices/PDA/PDA.dm:928]:
```suggestion
		target.AddComponent(/datum/component/sound_player, amount = (rand(15,20)), signal_or_sig_list = sig_list)
		installed_cartridge.charges --
		return TRUE

```
You'd want to stop the attack chain and not slap or open the airlock after this is done, correct?

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 8 above)
Comment (KathyRyals) [reply code/game/objects/items/devices/PDA/PDA.dm:918]: Only if the door panel is open and you click on it with the PDA in your hand.

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 9 above)
Comment (KathyRyals) [reply code/game/objects/items/devices/PDA/PDA.dm:928]: Yep ! It doesn't slap currently though if I'm not mistaken

Comment (KathyRyals) [reply code/game/objects/items/devices/PDA/PDA.dm:928]: But I'll commit because Rohesie code good

File: code/datums/components/soundplayer.dm [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+	If amount_left is equal to -1, the component is infinite and will never delete itself.
+*/
+/datum/component/sound_player/proc/play_sound()
+	if(amount_left == -1)
+		playsound(parent, pickweight(sounds), volume, TRUE)
+		return
+	playsound(parent, pickweight(sounds), volume, TRUE)
+	amount_left --
+	if(amount_left == 0)
+		qdel(src)
```
Comment (Rohesie) [inline code/datums/components/soundplayer.dm:39]:
```suggestion
/datum/component/sound_player/proc/play_sound()
	playsound(parent, pickweight(sounds), volume, TRUE)
	switch(amount_left)
		if(-1)
			return
		if(1) //Last use.
			qdel(src)
			return
		else
			amount_left --
```

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 9 above)
Comment (Rohesie) [reply code/game/objects/items/devices/PDA/PDA.dm]: Doesn't it open it, if you have access?

File: code/datums/components/soundplayer.dm (hunk 10 above)
Comment (KathyRyals) [reply code/datums/components/soundplayer.dm:39]: Good code good

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 9 above)
Comment (KathyRyals) [reply code/game/objects/items/devices/PDA/PDA.dm]: Errhh lemme check

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 11]
```diff
@@ -904,6 +904,31 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+	to_chat(user, "<span class='notice'>You upload the virus to the airlock controller!</span>")
+	var/sig_list
+	if(istype(target,/obj/machinery/door/airlock))
+		sig_list += list(COMSIG_AIRLOCK_OPEN, COMSIG_AIRLOCK_CLOSE)
+	else
+		sig_list += list(COMSIG_ATOM_ATTACK_HAND)
+		//TODO : Add more signals for specific cases like panel open, printing, dispensing etc
+	target.AddComponent(/datum/component/sound_player, amount = (rand(15,20)), signal_or_sig_list = sig_list)
+	installed_cartridge.charges --
+	return ..()
```
Comment (Rohesie) [inline code/game/objects/items/devices/PDA/PDA.dm:929]:
```suggestion
	return TRUE
```
This is when we want it to stop with the attack.

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 12]
```diff
@@ -904,6 +904,31 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+	if(installed_cartridge.charges <=0)
+		to_chat(user, "<span class='notice'>Out of charges.</span>")
+		return ..()
+	to_chat(user, "<span class='notice'>You upload the virus to the airlock controller!</span>")
+	var/sig_list
+	if(istype(target,/obj/machinery/door/airlock))
+		sig_list += list(COMSIG_AIRLOCK_OPEN, COMSIG_AIRLOCK_CLOSE)
+	else
+		sig_list += list(COMSIG_ATOM_ATTACK_HAND)
+		//TODO : Add more signals for specific cases like panel open, printing, dispensing etc
```
Comment (Rohesie) [inline code/game/objects/items/devices/PDA/PDA.dm:926]:
```suggestion
```

**Review by Rohesie — APPROVED** [review]

**Pull request merged**